src/java/org/apache/cassandra/repair/messages/RepairMessage.java

```diff
@@ -42,57 +55,182 @@
+            }
+
+            @Override
             public void onFailure(InetAddressAndPort from, RequestFailureReason failureReason)
             {
-                logger.error("[#{}] {} failed on {}: {}", request.desc.parentSessionId, verb, from, failureReason);
+                failureCallback.onFailure(RepairException.error(request.desc, PreviewKind.NONE, String.format("Got %s failure from %s: %s", verb, from, failureReason)));
+            }
 
-                if (supportsTimeouts(from, request.desc.parentSessionId))
-                    failureCallback.onFailure(RepairException.error(request.desc, PreviewKind.NONE, String.format("Got %s failure from %s: %s", verb, from, failureReason)));
+            @Override
+            public boolean invokeOnFailure()
+            {
+                return true;
             }
         };
-
-        MessagingService.instance().sendWithCallback(Message.outWithFlag(verb, request, CALL_BACK_ON_FAILURE),
-                                                     endpoint,
-                                                     callback);
+        sendMessageWithRetries(ctx, allowRetry, request, verb, endpoint, callback);
     }
 
-    private static boolean supportsTimeouts(InetAddressAndPort from, TimeUUID parentSessionId)
+    private static ErrorHandling errorHandlingSupported(SharedContext ctx, InetAddressAndPort from, Verb verb, TimeUUID parentSessionId)
     {
-        CassandraVersion remoteVersion = Gossiper.instance.getReleaseVersion(from);
-        if (remoteVersion != null && remoteVersion.compareTo(SUPPORTS_TIMEOUTS) >= 0)
-            return true;
-        logger.warn("[#{}] Not failing repair due to remote host {} not supporting repair message timeouts (version = {})", parentSessionId, from, remoteVersion);
-        return false;
+        if (SUPPORTS_RETRY_WITHOUT_VERSION_CHECK.contains(verb))
+            return ErrorHandling.RETRY;
+        // Repair in mixed mode isn't fully supported, but also not activally blocked... so in the common case all participants
+        // will be on the same version as this instance, so can avoid the lookup from gossip
+        CassandraVersion remoteVersion = ctx.gossiper().getReleaseVersion(from);
+        if (remoteVersion == null)
+        {
+            if (VERB_TIMEOUT_VERSIONS.containsKey(verb))
+            {
+                logger.warn("[#{}] Not failing repair due to remote host {} not supporting repair message timeouts (version is unknown)", parentSessionId, from);
+                return ErrorHandling.NONE;
+            }
+            return ErrorHandling.TIMEOUT;
+        }
+        if (remoteVersion.compareTo(SUPPORTS_RETRY) >= 0)
+            return ErrorHandling.RETRY;
+        CassandraVersion timeoutVersion = VERB_TIMEOUT_VERSIONS.get(verb);
+        if (timeoutVersion == null ||
+            remoteVersion.compareTo(timeoutVersion) >= 0)
```

Comment:
nit: Might be nice for this to be on one line. I have to do a double-take w/ the `return` having the same indentation.